### DISCUSSION TITLE
Network usage

### DISCUSSION THREAD

**Discussion by b00f** [pactus-project/pactus#571] [General]
For some users who run Pactus on their personal computers or laptops, the cost of internet service can be high due to the monthly fee. This is compared to the one-time purchase of a CPU or storage device (Hard drive). Therefore, it's important to keep network usage as low as possible without harming the overall consensus mechanism or the operation of the blockchain.

At present, we don't have a really good measure of network usage. However, there are two interesting metrics that we can look at:

1- How many bytes the sync module [has sent or received](https://github.com/pactus-project/pactus/blob/2dd2f3bb62243fb58e347d5a891a1597744ee40e/sync/peerset/peer_set.go#L24).
2- If we run Pactus on Docker, we can see how many bytes a Docker instance [has sent or received](https://docs.docker.com/engine/reference/commandline/stats/).

Comparing these two can provide us with interesting insights. For example, in one of the public nodes, we have these metrics from the sync module:

```
Total Sent Bytes:       602434591
Total Received Bytes:   1196449061
```
Over a period of more than 3 days, we received 1.1 GB and sent 0.6 GB data.

Meanwhile, Docker provides us with these stats:

```
NET I/O 
69.7GB / 75.1GB
```

This means that during the same period, we received 69.7 GB and sent 75.1 GB.

This finding is quite interesting. It shows that Libp2p received 60 times more data than what we actually consumed. As for sending data, we sent 120 times more than what we actually sent out. I think that due to the nature of P2P networks, we may receive some data (like we are bridging data?) but this is a big difference.

So, what do we need to do?

## Gather More Metrics

In LibP2p, we can extract additional [metrics](https://github.com/libp2p/go-libp2p/blob/master/p2p/host/resource-manager/README.md) from the Resource Manager module. This would really help us get a better understanding of network usage in the Libp2p layer.

In our sync module, we need to detail exactly how much data we have sent or received per [message type](https://github.com/pactus-project/pactus/blob/2dd2f3bb62243fb58e347d5a891a1597744ee40e/sync/bundle/message/message.go#L42).

## Limit Inbound/Outbound Rates

We can define some configurable parameters that allow users to set limits for the inbound and outbound rates. 

## Reduce Network Usage in the Sync Module

We can reduce data usage without affecting the consensus mechanism. For instance, each node can stop broadcasting the  [block-announce message](https://github.com/pactus-project/pactus/blob/main/sync/bundle/message/block_announce.go) if it's received from another peer. In this case, at each height, instead of 21 block-announce messages, we might have fewer than 10. There may also be other ways to reduce the network usage.

## Reduce Network Usage in the Libp2p Layer

We should explore ways to do this.

We can try data compression in Libp2p. I'm not sure if compressing data would help us, especially as most of the data is random, like hashes or signatures. It seems that compressing data might not give us a significant benefit. However, it's worth noting that we already have our own [compression](https://github.com/pactus-project/pactus/blob/2dd2f3bb62243fb58e347d5a891a1597744ee40e/sync/handler_blocks_response.go#L49) for block-response messages.

We could also limit the number of open connections, but we need to determine the optimal and secure.

Is there any thing else, we can do?

--- 
Some useful links:

- Interesting way to compact blocks by removing transaction from most recent blocks in Bitcoin:  https://bitcoincore.org/en/2016/06/07/compact-blocks-faq/
- Get peer info API in bitcoin: https://developer.bitcoin.org/reference/rpc/getpeerinfo.html
- Compression in Libp2p: https://discuss.libp2p.io/t/compression-in-libp2p/728
- DoS Mitigation in Libp2p:  https://docs.libp2p.io/concepts/security/dos-mitigation/

**Comment by b00f** [comment]
@kehiy @amirvalhalla

**Reply by amirvalhalla** [reply]
@b00f 

I think for the first step it would be better to add some metrics to `Pactus`, this way, we could determine the usage of different sections of code better then we can consider what parts of the code need to compress or use other solutions.

if you agree to this solution, first we should add metrics after that consider the result of the metrics then we should think about the solutions , and I'll try to write metrics for `Pactus`

besides all of that, we should talk together to decide which sections need metrics.

**Reply by kehiy** [reply]
@b00f @amirvalhalla 
This is a complex and important issue that we should be careful
about for now, I need more research to get my ideas, but there is some stuff I can provide for now!
so, first of all, we know that more metrics and more detailed metrics can help us to find the best way to optimize network and bandwidth usage

the first way to get more detailed metrics is by adding a `GRPC`or `HTTP` endpoint as `EXPORTER` that gets all metrics of the network with one API call. (we should add specific parts of network usage and...) 

with this action, we can do some important stuff:
1. use this endpoint for some monitoring tools like [prometheus](https://prometheus.io/docs/prometheus/latest/getting_started/) it helps us to analyze metrics with a good visualized environment and get users a good option to know their node status easily and detailed (most blockchains full node owners like file coin or.. that use aws or other things for this are using this way for monitoring)

2. but the previous thing is good for Visual inspection and analysis, but the  `EXPORTER` endpoint itself because it returns a `JSON`, `PorotoBuf`, `XML`, or... we can use it for machine learning or data since analysis too... (I mean all analysis by machines and codes)

this two-step is all about getting metrics better
To reduce the network usage we need more research (especially for me as a newbie), so I will read the codes, and resources @b00f provided till tomorrow and I will share my ideas for it!

**Reply by b00f** [reply]
I think we should arrange a DEV-meeting for this.

**Reply by amirvalhalla** [reply]
@b00f 
keep us updated on DEV-meeting time

**Reply by kehiy** [reply]
good idea!
when we can have the DEV meeting?

**Reply by b00f** [reply]
This Saturday? Or next week?

**Reply by amirvalhalla** [reply]
@b00f 
I'm okay with both

**Reply by kehiy** [reply]
@b00f Saturday that better, but I'm ok  with both too

**Reply by b00f** [reply]
Ok, Let's fix it on Saturday evening. Tomorrow we work on the announcement together.

**Comment by fonty1** [comment]
Hey, very interesting thread! 

Have you considered using a decentralized publish/subscribe network to relay your blocks and other related consensus information?

One such option is the [Streamr Network](https://docs.streamr.network/streamr-network/what-is-streamr-network). The transport protocol is WebRTC so peers can connect through difficult NATs, it has very low and predictable latency, it's decentralized, permissionless and so on. Your nodes will connect to 4 nodes in your topology to relay messages, so its much more efficient than Gossip flooding. 
[Whitepaper](https://streamr-public.s3.amazonaws.com/streamr-network-scalability-whitepaper-2020-08-20.pdf)

Happy to answer any questions about the protocol. No tokens are needed. We also do decentralized metrics gathering if you'd like to better inspect your nodes in a decentralized way. 
[Polygon case study](https://streamr.network/hub/projects/0xd0f8ff509ce2581af1be159bdc1f704cc7a2b8202e594fa902624588dea0a554/overview)
[Blog article on decentralized metrics gathering](https://blog.streamr.network/collecting-metrics-data-from-decentralized-systems/)

**Reply by amirvalhalla** [reply]
@fonty1 

thanks for your suggestion, we have a `DEV-meeting` which will be held very soon, we'll consider your offer and talk about it, also you could join our `DEV-meeting` and there we discuss different solutions.
[Our Discord Server](https://discord.gg/e7NtzjhzDZ)
[Dev Meeting Announcement](https://discord.com/channels/795592769300987944/811878389304655892/1126835812122763264)

**Reply by fonty1** [reply]
Thanks @amirvalhalla - I've joined your Discord server and I'll try to make it to the call this Sunday.

**Reply by amirvalhalla** [reply]
@fonty1 
great, it's an honor for us to have you at the meeting.
